File: java/Samples/Biometrics/VB/AbisSampleVB/SubjectEditor/TenPrintCard/TenPrintCardPrintForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Drawing.Printing
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Images

Partial Public Class TenPrintCardPrintForm
	Inherits PrintPreviewDialog
#Region "Private fields"

	Private ReadOnly _rolledPositions() As NFPosition = {NFPosition.RightThumb, NFPosition.RightIndex, NFPosition.RightMiddle, NFPosition.RightRing, NFPosition.RightLittle, NFPosition.LeftThumb, NFPosition.LeftIndex, NFPosition.LeftMiddle, NFPosition.LeftRing, NFPosition.LeftLittle}

	Private _painter As FramePainter
	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _images As Dictionary(Of Integer, NImage)

#End Region

#Region "Public constructor"

	Public Sub New(ByVal frameDefinition As String)
		InitializeComponent()
		_painter = New FramePainter(frameDefinition)
		_images = New Dictionary(Of Integer, NImage)()
	End Sub

#End Region

#Region "Public properties"

	Public Property Subject() As NSubject
		Get
			Return _subject
		End Get
		Set(ByVal value As NSubject)
			_subject = value
		End Set
	End Property

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Events"

	Private Sub TenPrintCardPrintFormShown(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Shown
		Dim doc = New PrintDocument()
		AddHandler doc.PrintPage, AddressOf PrintPreviewEvent
		AddHandler doc.BeginPrint, AddressOf BeginPrintEvent
		Document = doc
	End Sub

	Private Sub PrintPreviewEvent(ByVal o As Object, ByVal e As PrintPageEventArgs)
		Dim graphics = e.Graphics
		Dim rect = _painter.GetFrameSplitting()

		graphics.TranslateTransform(13, 25)
		If _subject.Fingers.Count > 0 Then
			graphics.CompositingQuality = CompositingQuality.HighQuality
			graphics.SmoothingMode = SmoothingMode.HighQuality
			graphics.InterpolationMode = InterpolationMode.HighQualityBicubic

			Dim subject = MakeSubject()

			SegmentFingers(subject)

			For i As Integer = 1 To 14
				If _images.ContainsKey(i) Then
					Dim currrentRect = rect(i)
					Dim image = _images(i).ToBitmap()
					Dim tmpRect = New Rectangle(currrentRect.X + 2, currrentRect.Y + 2, currrentRect.Width - 4, currrentRect.Height - 4)

					Dim imagePosition = Me.ImagePosition(tmpRect, image)
					graphics.DrawImage(image, imagePosition)
				End If
			Next i
		End If

		_painter.DrawForm(graphics)

	End Sub

	Private Sub BeginPrintEvent(ByVal sender As Object, ByVal e As PrintEventArgs)
		Document.DefaultPageSettings.PrinterResolution = New PrinterResolution With {.Kind = PrinterResolutionKind.Custom, .X = 100, .Y = 100}
		Document.DefaultPageSettings.PaperSize = Document.PrinterSettings.PaperSizes.Cast(Of PaperSize)().Where(Function(x) x.Kind = PaperKind.A4).First()
		If (Not Document.PrintController.IsPreview) Then
			PrintDialog1.Document = Document
			PrintDialog1.AllowSelection = False
			PrintDialog1.AllowSomePages = False
			If PrintDialog1.ShowDialog() <> DialogResult.OK Then
				e.Cancel = True
			End If
		End If
	End Sub

#End Region

#Region "Private methods"

	Private Function MakeSubject() As NSubject
		Dim subject = New NSubject()
		Dim fingers = _subject.Fingers.Where(Function(x) x.Image IsNot Nothing)
		Dim rolled = fingers.Where(Function(x) NBiometricTypes.IsImpressionTypeRolled(x.ImpressionType)).Where(Function(x) NBiometricTypes.IsPositionOneOf(x.Position, _rolledPositions)).Select(Function(x) New NFinger With {.Image = x.Image, .Position = x.Position, .ImpressionType = x.ImpressionType}).GroupBy(Function(x) x.Position).Select(Function(x) x.First())

		For Each finger In rolled
			subject.Fingers.Add(finger)
		Next finger

		Dim thumbPositions = New NFPosition() {NFPosition.LeftThumb, NFPosition.RightThumb, NFPosition.PlainRightThumb, NFPosition.PlainLeftThumb}
		Dim plain = fingers.Where(Function(x) NBiometricTypes.IsImpressionTypePlain(x.ImpressionType)).Where(Function(x) NBiometricTypes.IsPositionFourFingers(x.Position) OrElse NBiometricTypes.IsPositionOneOf(x.Position, thumbPositions)).Select(Function(x) New NFinger With {.Image = x.Image, .Position = x.Position, .ImpressionType = x.ImpressionType}).GroupBy(Function(x) x.Position).Select(Function(x) x.First())

		For Each finger In plain
			subject.Fingers.Add(finger)
		Next finger

		Return subject
	End Function

	Private Sub SegmentFingers(ByVal subject As NSubject)
		_images.Clear()
		Using task = _biometricClient.CreateTask(NBiometricOperations.Segment, subject)
			_biometricClient.PerformTask(task)
			If task.Status <> NBiometricStatus.Ok Then
				Throw New Exception("Segmentation failed")
			End If

			Dim segmentedFingers = subject.Fingers.Where(Function(x) x.ParentObject IsNot Nothing)
			Dim rolled = segmentedFingers.Where(Function(x) NBiometricTypes.IsImpressionTypeRolled(x.ImpressionType))

			Dim rolledPosition = _rolledPositions.ToList()
			For Each finger In rolled
				If rolledPosition.Contains(finger.Position) Then
					Dim index As Integer = rolledPosition.IndexOf(finger.Position) + 1
					_images.Add(index, finger.Image)
				End If
			Next finger

			Dim thumbs = segmentedFingers.Where(Function(x) NBiometricTypes.IsImpressionTypePlain(x.ImpressionType)).Where(Function(x) NBiometricTypes.IsPositionSingleFinger(x.Position) AndAlso x.Position.ToString().Contains("Thumb"))
			Dim leftThumb = thumbs.Where(Function(x) NBiometricTypes.IsPositionLeft(x.Position)).FirstOrDefault()
			If leftThumb IsNot Nothing Then
				_images.Add(12, leftThumb.Image)
			End If

			Dim rightThumb = thumbs.Where(Function(x) NBiometricTypes.IsPositionRight(x.Position)).FirstOrDefault()
			If rightThumb IsNot Nothing Then
				_images.Add(13, rightThumb.Image)
			End If

			Dim fourFingers = subject.Fingers.Where(Function(x) x.ParentObject Is Nothing AndAlso NBiometricTypes.IsPositionFourFingers(x.Position))
			Dim leftFour = fourFingers.Where(Function(x) NBiometricTypes.IsPositionLeft(x.Position)).FirstOrDefault()
			If leftFour IsNot Nothing Then
				_images.Add(11, GetFingerImage(leftFour))
			End If

			Dim rightFour = fourFingers.Where(Function(x) NBiometricTypes.IsPositionRight(x.Position)).FirstOrDefault()
			If rightFour IsNot Nothing Then
				_images.Add(14, GetFingerImage(rightFour))
			End If
		End Using
	End Sub

	Private Function GetFingerImage(ByVal finger As NFinger) As NImage
		Dim rect As Rectangle = finger.Objects.First().BoundingRect
		For i As Integer = 1 To finger.Objects.Count - 1
			rect = Rectangle.Union(rect, finger.Objects(i).BoundingRect)
		Next i

		Dim fingerImage = finger.Image
		If rect.X < 0 Then
			rect.X = 0
		End If
		If rect.Y < 0 Then
			rect.Y = 0
		End If
		If rect.Width + rect.X > fingerImage.Width Then
			rect.Width = CInt(Fix(fingerImage.Width)) - rect.X
		End If
		If rect.Height + rect.Y > fingerImage.Height Then
			rect.Height = CInt(Fix(fingerImage.Height)) - rect.Y
		End If

		Return finger.Image.Crop(CUInt(rect.X), CUInt(rect.Y), CUInt(rect.Width), CUInt(rect.Height))
	End Function

	Private Function ImagePosition(ByVal rect As Rectangle, ByVal image As Image) As Rectangle
		Dim widthRatio As Double = 1.0 * rect.Width / image.Width
		Dim heightRatio As Double = 1.0 * rect.Height / image.Height

		Dim ratio As Double = If(widthRatio < heightRatio, widthRatio, heightRatio)
		Dim width As Integer = CInt(Fix(Math.Truncate(image.Width * ratio)))
		Dim height As Integer = CInt(Fix(Math.Truncate(image.Height * ratio)))

		Dim rectangle = New Rectangle(rect.X, rect.Y, width, height)
		Dim centerWidth As Integer = rect.Width - width
		If centerWidth > 0 Then
			rectangle.X += centerWidth / 2
		End If

		Dim centerHeight As Integer = rect.Height - height
		If centerHeight > 0 Then
			rectangle.Y += centerHeight / 2
		End If

		Return rectangle
	End Function

#End Region

End Class

File: java/Samples/Biometrics/CS/AbisSampleCS/SubjectEditor/TenPrintCard/TenPrintCardPrintForm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Printing;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Images;

namespace Neurotec.Samples.SubjectEditor.TenPrintCard
{
	public partial class TenPrintCardPrintForm : PrintPreviewDialog
	{
		#region Private fields

		private readonly NFPosition[] _rolledPositions =
				{
					NFPosition.RightThumb, NFPosition.RightIndex, NFPosition.RightMiddle, NFPosition.RightRing, NFPosition.RightLittle,
					NFPosition.LeftThumb, NFPosition.LeftIndex, NFPosition.LeftMiddle, NFPosition.LeftRing, NFPosition.LeftLittle
				};

		private FramePainter _painter;
		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private Dictionary<int, NImage> _images;

		#endregion

		#region Public constructor

		public TenPrintCardPrintForm(string frameDefinition)
		{
			InitializeComponent();
			_painter = new FramePainter(frameDefinition);
			_images = new Dictionary<int, NImage>();
		}

		#endregion

		#region Public properties

		public NSubject Subject
		{
			get { return _subject; }
			set { _subject = value; }
		}

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set { _biometricClient = value; }
		}

		#endregion

		#region Events

		private void TenPrintCardPrintFormShown(object sender, EventArgs e)
		{
			var doc = new PrintDocument();
			doc.PrintPage += PrintPreviewEvent;
			doc.BeginPrint += BeginPrintEvent;
			Document = doc;
		}

		private void PrintPreviewEvent(object o, PrintPageEventArgs e)
		{
			var graphics = e.Graphics;
			var rect = _painter.GetFrameSplitting();

			graphics.TranslateTransform(13, 25);
			if (_subject.Fingers.Count > 0)
			{
				graphics.CompositingQuality = CompositingQuality.HighQuality;
				graphics.SmoothingMode = SmoothingMode.HighQuality;
				graphics.InterpolationMode = InterpolationMode.HighQualityBicubic;

				var subject = MakeSubject();

				SegmentFingers(subject);

				for (int i = 1; i <= 14; i++)
				{
					if (_images.ContainsKey(i))
					{
						var currrentRect = rect[i];
						var image = _images[i].ToBitmap();
						var tmpRect = new Rectangle(currrentRect.X + 2, currrentRect.Y + 2, currrentRect.Width - 4, currrentRect.Height - 4);

						var imagePosition = ImagePosition(tmpRect, image);
						graphics.DrawImage(image, imagePosition);
					}
				}
			}

			_painter.DrawForm(graphics);

		}

		private void BeginPrintEvent(object sender, PrintEventArgs e)
		{
			Document.DefaultPageSettings.PrinterResolution = new PrinterResolution { Kind = PrinterResolutionKind.Custom, X = 100, Y = 100 };
			Document.DefaultPageSettings.PaperSize = Document.PrinterSettings.PaperSizes.Cast<PaperSize>().Where(x => x.Kind == PaperKind.A4).First();
			if (!Document.PrintController.IsPreview)
			{
				printDialog1.Document = Document;
				printDialog1.AllowSelection = false;
				printDialog1.AllowSomePages = false;
				if (printDialog1.ShowDialog() != DialogResult.OK)
				{
					e.Cancel = true;
				}
			}
		}

		#endregion

		#region Private methods

		private NSubject MakeSubject()
		{
			var subject = new NSubject();
			var fingers = _subject.Fingers.Where(x => x.Image != null);
			var rolled = fingers.Where(x => NBiometricTypes.IsImpressionTypeRolled(x.ImpressionType))
				.Where(x => NBiometricTypes.IsPositionOneOf(x.Position, _rolledPositions))
				.Select(x => new NFinger { Image = x.Image, Position = x.Position, ImpressionType = x.ImpressionType })
				.GroupBy(x => x.Position)
				.Select(x => x.First());

			foreach (var finger in rolled)
			{
				subject.Fingers.Add(finger);
			}

			var thumbPositions = new NFPosition[] { NFPosition.LeftThumb, NFPosition.RightThumb, NFPosition.PlainRightThumb, NFPosition.PlainLeftThumb };
			var plain = fingers.Where(x => NBiometricTypes.IsImpressionTypePlain(x.ImpressionType))
				.Where(x => NBiometricTypes.IsPositionFourFingers(x.Position) || NBiometricTypes.IsPositionOneOf(x.Position, thumbPositions))
				.Select(x => new NFinger { Image = x.Image, Position = x.Position, ImpressionType = x.ImpressionType })
				.GroupBy(x => x.Position)
				.Select(x => x.First());

			foreach (var finger in plain)
			{
				subject.Fingers.Add(finger);
			}

			return subject;
		}

		private void SegmentFingers(NSubject subject)
		{
			_images.Clear();
			using (var task = _biometricClient.CreateTask(NBiometricOperations.Segment, subject))
			{
				_biometricClient.PerformTask(task);
				if (task.Status != NBiometricStatus.Ok)
				{
					throw new Exception("Segmentation failed");
				}

				var segmentedFingers = subject.Fingers.Where(x => x.ParentObject != null);
				var rolled = segmentedFingers.Where(x => NBiometricTypes.IsImpressionTypeRolled(x.ImpressionType));

				var rolledPosition = _rolledPositions.ToList();
				foreach (var finger in rolled)
				{
					if (rolledPosition.Contains(finger.Position))
					{
						int index = rolledPosition.IndexOf(finger.Position) + 1;
						_images.Add(index, finger.Image);
					}
				}

				var thumbs = segmentedFingers.Where(x => NBiometricTypes.IsImpressionTypePlain(x.ImpressionType))
					.Where(x => NBiometricTypes.IsPositionSingleFinger(x.Position) && x.Position.ToString().Contains("Thumb"));
				var leftThumb = thumbs.Where(x => NBiometricTypes.IsPositionLeft(x.Position)).FirstOrDefault();
				if (leftThumb != null)
				{
					_images.Add(12, leftThumb.Image);
				}

				var rightThumb = thumbs.Where(x => NBiometricTypes.IsPositionRight(x.Position)).FirstOrDefault();
				if (rightThumb != null)
				{
					_images.Add(13, rightThumb.Image);
				}

				var fourFingers = subject.Fingers.Where(x => x.ParentObject == null && NBiometricTypes.IsPositionFourFingers(x.Position));
				var leftFour = fourFingers.Where(x => NBiometricTypes.IsPositionLeft(x.Position)).FirstOrDefault();
				if (leftFour != null)
				{
					_images.Add(11, GetFingerImage(leftFour));
				}

				var rightFour = fourFingers.Where(x => NBiometricTypes.IsPositionRight(x.Position)).FirstOrDefault();
				if (rightFour != null)
				{
					_images.Add(14, GetFingerImage(rightFour));
				}
			}
		}

		private NImage GetFingerImage(NFinger finger)
		{
			Rectangle rect = finger.Objects.First().BoundingRect;
			for (int i = 1; i < finger.Objects.Count; i++)
			{
				rect = Rectangle.Union(rect, finger.Objects[i].BoundingRect);
			}

			var fingerImage = finger.Image;
			if (rect.X < 0) rect.X = 0;
			if (rect.Y < 0) rect.Y = 0;
			if (rect.Width + rect.X > fingerImage.Width)
				rect.Width = (int)fingerImage.Width - rect.X;
			if (rect.Height + rect.Y > fingerImage.Height)
				rect.Height = (int)fingerImage.Height - rect.Y;

			return finger.Image.Crop((uint)rect.X, (uint)rect.Y, (uint)rect.Width, (uint)rect.Height);
		}

		private Rectangle ImagePosition(Rectangle rect, Image image)
		{
			double widthRatio = 1.0 * rect.Width / image.Width;
			double heightRatio = 1.0 * rect.Height / image.Height;

			double ratio = widthRatio < heightRatio ? widthRatio : heightRatio;
			int width = (int)Math.Truncate(image.Width * ratio);
			int height = (int)Math.Truncate(image.Height * ratio);

			var rectangle = new Rectangle(rect.X, rect.Y, width, height);
			int centerWidth = rect.Width - width;
			if (centerWidth > 0)
			{
				rectangle.X += centerWidth / 2;
			}

			int centerHeight = rect.Height - height;
			if (centerHeight > 0)
			{
				rectangle.Y += centerHeight / 2;
			}

			return rectangle;
		}

		#endregion

	}
}
